### PR TITLE
identify item need

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
## Summary
Sorting items by how soon the user might need to buy them will make it easier for the user to identify what they need to buy, which will in turn help them better plan their shopping trips. To that end, we will be sorting items into 4 possible groups, according to how urgently the user should buy the item:

1. Need to buy soon (7 days or fewer until the next purchase)
2. Need to buy kind of soon (between 7 & 30 days until the next purchase)
3. Need to buy not soon (30 days or more until the next purchase)
4. Inactive (60 days have passed since the last purchase)

To implement this feature, your team will need to create some new UI elements so the user knows how their items are sorted and write some logic behind-the-scenes to perform the sorting itself. Read the acceptance criteria and our notes carefully to be sure you understand the tasks at hand, and don’t be afraid to reach out to the rest of your team, and your mentors if you need help. You’ve got this! 

## Acceptance criteria
- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically

### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Notes
There's a lot to do for this issue! Read this section for some ideas about how to approach the tasks ahead. The `getDaysBetweenDates` function in `src/utils/dates.js` will be useful!

Remember: The dates in the Firestore database are special `Timestamp` objects. You’ll need to call their `toDate()` method to get a JavaScript Date that we can work with.

### Indicating purchase urgency in the UI
It’s tempting to use colors or icons to indicate how soon the user should buy an item. Keep in mind:
- If you use only color, a colorblind user won’t be able to tell the difference between urgency indicators
- If you use icons, you have to teach users what those icons mean

With these constraints, it’s probably best to render plain text, like “soon”, “kind of soon”, etc. in order to show urgency. You can still use color alongside that text, and the design of where that text goes is up to you!

### Sorting the list
We’re sorting by multiple criteria:
1. Whether the item is active or inactive, then
2. By the number of days until its `dateNextPurchased`, and
3. Alphabetically by name, if they have the same number of days until purchase

Sorting by multiple criteria can be tricky! Refer to the MDN docs for [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and refresh yourself on how to write *compare functions* so you can write the `comparePurchaseUrgency` function.

### PR DESCRIPTION
**Pull request by ToobaJamal**
Co-authored-by: @shagunZ 

## Description
The items are sorted alphabetically making it easier for the user to identify which items they need to purchase according to the urgency. The item urgency is shown with an indicator as soon, kind of soon, not soon, and inactive.
If an item’s dateNextPurchased has passed and it is not yet inactive, the item will be shown with an "overdue" label.

A function `comparePurchaseUrgency` is added to `firebase.js` which compares the purchase urgency of items and sorts them according to the requirements.

## Related Issue

Closes #12 

## Acceptance Criteria

- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
 - [x] This urgency indicator does not rely on only color
 - [x] api/firestore.js exports a new comparePurchaseUrgency function with the following behaviors
   - [x] sorts inactive items last, then
   - [x] sorts items in ascending order of days until purchase, and
   - [x] sorts items with the same days until purchase alphabetically
### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s dateNextPurchased has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

 - [x] Extend the functionality of comparePurchaseUrgency to sort “overdue” items to the top of the list
 - [x] Indicate in your UI when an item is overdue

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

Screenshot (104)

### After

https://github.com/the-collab-lab/tcl-60-smart-shopping-list/assets/100477248/e91403c9-ff95-4d27-9af4-5cc162d78f36

## Testing Steps / QA Criteria

Open a shopping list in your browser and verify that the urgency indicator is displayed with each item in the list.

### CODE REVIEW

File: src/utils/dates.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ export function getFutureDate(offset) {
 }
 
 export function getDaysBetweenDates(date1, date2) {
-	const diffInMs = Math.abs(date1 - date2);
```
Comment (RajGM) [inline src/utils/dates.js:15]: Why removed the absolution function?

File: src/api/firebase.js [hunk 2]
```diff
@@ -126,3 +128,37 @@ export async function checkItem(listId) {
...
+		if (item.dateNextPurchased != null) {
+			let daysDiff = getDaysBetweenDates(
+				item.dateNextPurchased.toDate().getTime(),
+				new Date().getTime(),
+			);
+			item.urgency =
+				daysDiff >= 30
+					? 'Not soon'
+					: daysDiff >= 7
+					? 'Kind of Soon'
```
Comment (RajGM) [inline src/api/firebase.js]: Brackets around the nested ternary operator will make the code more readable

**Review by labrocadabro — COMMENTED** [review]
Looks good! A few small notes:

1. I'm not sure firebase.js was the best place to add this code. It doesn't seem to be related to firebase? Maybe could have put it a new file, something like `utils/urgency.js`? And then just import the function to call it.

2. Considering all the properties of `item` we're now passing from List to ListItem, we could probably just spread it.

3. I'm curious why `Math.abs` was removed from `const diffInMs = Math.abs(date1 - date2);` in dates.js? And why was it there in the first place? Did removing it break anything that was relying on it?

**Review by RajGM — APPROVED** [review]
Looks good to me.

**Pull request merged**